ENH: plot adjacency

mmagnuski opened the pull request:
#### Reference issue
Fixes #10602.


#### What does this implement/fix?
This adds `plot_ch_adjacency`. Currently this is a clean-up, fixed version of what I had in my `sarna` repo. In matplotlib interactive mode it allows to edit the adjacency too:
![plot_and_edit_adjacency](https://user-images.githubusercontent.com/8452354/168424543-62f1fbd3-cbaa-4dce-ac86-db3f19a6b55c.gif)
Currently a bit more work is needed, see below:

#### TODOs:
- [x] rename to `plot_ch_adjacency`
- [x] 3d mode (`kind='3d'`) should turn off edit mode as poistion of mouse clicks are not reliable in three dimensions in matplotlib
- [x] currently if adjacency is sparse it is turned to array. This makes sense for convenient visualisation, but makes inplace modifications to the adjacency impossible (now a warning is raised, but this could be changed into an error when if we add a default of `enable_edit=False`.
- [x] scale node size with number of connections
- [x] change node size as number of connection changes during user interaction
- [x] remove toolbar
- [x] it should likely do something about channel types, now it just assumes all channels are of the same type
- [x] it should additionally test if adjacency shape matches the number of channels available in the passed mne object.
- [x] tests
  - [x] basic 2d plotting
  - [x] 2d plotting interactivity
  - [x] 3d mode
  - [x] errors thrown
- [x] whats new
- [x] some example?

#### How to run
```python
import mne
from mne.datasets import sample

data_path = sample.data_path()
meg_path = data_path / 'MEG' / 'sample'
raw_fname = meg_path / 'sample_audvis_filt-0-40_raw.fif'

raw = mne.io.read_raw_fif(raw_fname).crop(0, 5).pick('eeg').load_data()

# %%
adj, ch_names = mne.channels.find_ch_adjacency(raw.info, 'eeg')

# %%
mne.viz.topomap.plot_ch_adjacency(raw, adj, kind='2d')
```
to make the edits work in place, do `adj = adj.toarray()` before plotting adjacency.

mmagnuski commented:
for more context: I have found the option to manually edit the adjacency very useful in the past. The deafault triangularization done by mne often requires some manual corrections in my experience.

hoechenberger commented:
omg this is amazing!!!!

I never thought about "fixing" things manually. Could you give an example where you believe this might be useful?

mmagnuski commented:
I can look for some better examples but the changes I apply in the gif I pasted make sense (would have to confirm in 3d). BTW this is also what FieldTrip allows you to do.

sappelhoff commented:
This is off to an amazing start and more than I even wished for with the editing! Will look in more detail soon

hoechenberger commented:
@mmagnuski Thank you, it appears the animation didn't load properly for me first, hence I was confused :) But now I understand what you mean, and I have to say I'm almost slightly shocked as I realize I never thought of this problem before. I always just assumed that MNE would do "the right thing", but it's clear that it's **not** so clear what exactly the "right thing" would be in certain situations.

I'm wondering if `find_ch_adjacency()` should spit out a warning if the distance of two "adjacent" channels exceeds a certain threshold – this would help discover at least some of the problematic cases, I suppose?

Regarding your visualization:
Could we make it such that the size of each node is determined by the number of its edges / neighbors? This would make it easier to spot "well-connected" nodes and those who are rather "isolated".

Also I'm wondering if it could make sense to turn of the Matplotlib toolbar – I think all those buttons add unhelpful visual clutter – the only one I deem useful is the "Save" button, but it's not that we have one of those in all of our other visualizations either…

mmagnuski commented:
> I always just assumed that MNE would do "the right thing", but it's clear that it's not so clear what exactly the "right thing" would be in certain situations.

True, in many cases it wouldn't change the cluster based test results much, I think, but I've had clear cases where adjacency built by mne was problematic. I moved to mne from FieldTrip so I already had the habit to always visualize the adjacency and correct if it was needed. I can also reassure you that manual correction of adjacency is not a quirk :), see for example description of how FieldTrip adjacency templates are created (from [here](https://www.fieldtriptoolbox.org/template/neighbours/)):

> EEG neighbour templates are based on [automatic, symmetric triangulation](https://www.fieldtriptoolbox.org/faq/how_does_ft_prepare_neighbours_work) of the [2D layout templates](https://www.fieldtriptoolbox.org/template/layout). After the automatic definition, we used [3D electrode positions](https://www.fieldtriptoolbox.org/template/electrode) and [ft_neighbourplot](https://www.fieldtriptoolbox.org/reference/ft_neighbourplot) with cfg.enableedit=’yes’ to manually add and/or remove edges between sensors.

BTW, FieldTrip and mne don't construct the adjacency exactly the same - FieldTrip does a union of two triangulations ("symmetric triangulation" in the description above), mne does only one. As a conseqence you can get adjacency like the one posted by @sappelhoff in FieldTrip:
<img src=https://user-images.githubusercontent.com/8452354/168471297-395cccf9-fd41-4fc4-9c2a-4d710aa464c8.png width='650pix'>
but in mne you wouldn't get "crossed" adjacency relations (which in the case above clearly makes sense) as only one triangulation is performed. But all these adjacency-creation issues can be discussed in a separate issue thread, I think.

> Could we make it such that the size of each node is determined by the number of its edges / neighbors? This would make it easier to spot "well-connected" nodes and those which are rather "isolated".

Sure, I can add it. I never actually liked that option in FieldTrip (channels on the cap edges always have less connections and I didn't like that these nodes are drawn smaller :)), so maybe we could allow to turn it off with `scale_nodes=False` for example?

> Also I'm wondering if it could make sense to turn off the Matplotlib toolbar – I think all those buttons add unhelpful visual clutter – the only one I deem useful is the "Save" button, but it's not that we have one of those in all of our other visualizations either…

Definitely, I'll change that!

mmagnuski commented:
I can add `enable_edit=False` default so that no edits can be made unless user explicitly asks for this.

mmagnuski commented:
@hoechenberger I've changed node size (it scales with number of connections), now it looks like this:
<img src=https://user-images.githubusercontent.com/8452354/168476705-c7cfe74c-2fae-45ba-a639-88b5bcd130c2.gif width='500pix'>

hoechenberger commented:
Super nice!!

mmagnuski commented:
I've added some checks for `ch_type` but since I rarely had to handle data with multiple channel types, I'm not very good with handling types in mne, so please let me know how to do it better.

If everyone is ok with it, I will add `enable_edit=False` by default (this is a mirror of FieldTrips `enableedit`, but I can use a more mne'sqe name for that ) and option to disable node scaling (`scale_nodes`).
One thing I'm still thinking about is that `find_ch_adjacency` returns the adjacency and ch_names, so instead of checking for ch_type etc. maybe its better to require passing `ch_names` to `plot_ch_adjacency` and just do `inst.copy().pick_channels(ch_names, ordered=True)` and it will handle all the checks for me.

sappelhoff commented:
> If everyone is ok with it, I will add enable_edit=False by default (this is a mirror of FieldTrips enableedit, but I can use a more mne'sqe name for that ) and option to disable node scaling (scale_nodes).

sounds good!

regarding a potential example of this code in the docs, you could try to replace the plot here: https://mne.tools/stable/auto_tutorials/stats-sensor-space/75_cluster_ftest_spatiotemporal.html#find-the-fieldtrip-neighbor-definition-to-setup-sensor-adjacency with your new feature :thinking: 

unless you want to explicitly show the interactivity in a GIF or so.

> One thing I'm still thinking about is that find_ch_adjacency returns the adjacency and ch_names, so instead of checking for ch_type etc. maybe its better to require passing ch_names to plot_ch_adjacency and just do inst.copy().pick_channels(ch_names, ordered=True) and it will handle all the checks for me.

Is there ever a situation where one would want to plot adjacencies **across** channel types?

mmagnuski commented:
> unless you want to explicitly show the interactivity in a GIF or so.

Good, I'll update that example, and the gif and description interactivity options etc. I can add to a separate example in a next PR.

> Is there ever a situation where one would want to plot adjacencies across channel types?

I don't think so. But using the `ch_names` instead of `ch_type` should be both simpler (handles multiple checks "for free") and more stringent (makes sure channels are in the same order in the mne object and adjacency matrix). Also constructing adjacency already requires ch_type argument and returns ch_names - so the `ch_names` will be always from the same type.
I think I convinced myself, and I'll change the optional `ch_type` to requred `ch_names`. :)

hoechenberger commented:
I would suggest to call the parameter `edit` or `editable`, no? :)

sappelhoff commented:
> I don't think so. But using the ch_names instead of ch_type should be both simpler (handles multiple checks "for free") and more stringent (makes sure channels are in the same order in the mne object and adjacency matrix). Also constructing adjacency already requires ch_type argument and returns ch_names - so the ch_names will be always from the same type.
I think I convinced myself, and I'll change the optional ch_type to requred ch_names. :)

right, you also convinced me!

mmagnuski commented:
Ok, `edit` argument is now added, and the checks are simplified thanks to `ch_names`. I also got used to the node scaling, so I think I won't add an argument to turn it off. The only remaining things are tests (ouch), editing an example and whats new. I'll get to that tomorrow.

mmagnuski commented:
@larsoner Thanks, I know - its corrected now.
I've also already requested @sappelhoff review. :)

argh, and I forgot to correct one style issue...

sappelhoff commented on mne/viz/topomap.py:
```diff
@@ -2806,3 +2807,179 @@ def plot_bridged_electrodes(info, bridged_idx, ed_matrix, title=None,
+
+    Returns
+    -------
+    fig : matplotlib figure
+        Figure.
```
This fix should make the CIs pass

```suggestion
    fig : Figure
        The :class:`~matplotlib.figure.Figure` instance where the channel
        adjacency is plotted.
```

sappelhoff commented on mne/viz/utils.py:
```diff
@@ -2419,3 +2419,40 @@ def _handle_precompute(precompute):
+    Notes
+    -----
+    modified from:
+    http://stackoverflow.com/questions/13685386/matplotlib-equal-unit-length-with-equal-aspect-ratio-z-axis-is-not-equal-to
```
```suggestion
    https://stackoverflow.com/q/13685386
```

sappelhoff commented on mne/viz/topomap.py:
```diff
@@ -2806,3 +2807,180 @@ def plot_bridged_electrodes(info, bridged_idx, ed_matrix, title=None,
+        removes adjacency relation between the activated and newly clicked
+        channel (depending on whether the channels are already adjacent or
+        not); the newly clicked channel now becomes activated. Clicking on
+        an activated channel deactivates it.
+
```
```suggestion
        an activated channel deactivates it. Editing is currently only
        supported for ``kind='2d'``.

```

sappelhoff commented on mne/viz/topomap.py:
```diff
@@ -2806,3 +2807,180 @@ def plot_bridged_electrodes(info, bridged_idx, ed_matrix, title=None,
+        Info object with channel locations.
+    adjacency : array
+        Array of channels x channels shape. Defines which channels are adjacent
+        to each other.
```
```suggestion
        to each other. Note that if you edit adjacencies
        (via ``edit=True``), this array will be modified in place.
```

sappelhoff commented on mne/viz/tests/test_topomap.py:
```diff
@@ -728,3 +729,81 @@ def test_plot_bridged_electrodes():
+    adj = adj.toarray()
+
+    # plot adjacency
+    fig = plot_ch_adjacency(info, adj, ch_names, kind='2d', edit=True)
```
looks like you never call with `kind='3d'`, just doing this as a "smoke test" would be good, I think.

This would also improve coverage because your new `set_3d_axes_equal` would be called under the hood.

sappelhoff replied:
This would also improve coverage because your new `set_3d_axes_equal` would be called under the hood.

sappelhoff replied:
Ah I see this is still on your to do list! ... as well as including some tests that raise the expected errors, great!

sappelhoff commented on mne/viz/utils.py:
```diff
@@ -2419,3 +2419,40 @@ def _handle_precompute(precompute):
     return precompute
+
+
+def set_3d_axes_equal(ax):
```
I think I would make this private :thinking: ... because I would assume we remove this "helper" once MNE requires matplotlib > 3.3

sappelhoff reviewed (commented):
This is very high quality stuff, thanks @mmagnuski!

My open points are all covered by the items on your to do list! Please ping me if you want a final review.

I left a few minor comments.

sappelhoff commented:
@mmagnuski note also that I pushed a small fix, so remember to pull before continuing your work :wink:

mmagnuski commented:
Thanks @sappelhoff !
I will add the remaining tests today.

I don't get the error I'm getting on Circle:
```
Added  32 HTML plot_* redirects for auto_tutorials
Added  99 HTML plot_* redirects for auto_examples
Added  52 HTML custom redirects
Added "af386b75aa4a59e1766e64459fef97cde5160466" > _version.txt
make: *** [Makefile:70: html_dev-noplot] Error 1
make: Leaving directory '/home/circleci/project/doc'

Exited with code exit status 2
CircleCI received exit code 2
```

larsoner commented:
Looks like CircleCI is green. When you see that error, though, try searching for "warning" because we treat warnings as errors, as they usually indicate some problem with RST syntax

sappelhoff commented:
yeah, the circle error was a warning because you had "figure" as a type in a docstr, not "Figure" -- and only "Figure is defined as an alias in conf.py :-)

mmagnuski commented:
Ha, indeed, I refreshed the page and now its all green. @sappelhoff For some reason I saw the error in the last commit, after you fixed the figure problem.

mmagnuski commented:
@sappelhoff How do I pull these changes now? I looked online but I only found advice on how to create a PR or fetch someone else's PR...

mmagnuski commented:
Ok, I got it, I tried pulling from origin, not my fork, now it works.

sappelhoff commented:
Okay, great -- I would have thought that a simple `git pull` works :)

sappelhoff commented on mne/viz/topomap.py:
```diff
@@ -2806,3 +2807,182 @@ def plot_bridged_electrodes(info, bridged_idx, ed_matrix, title=None,
+    n_channels = len(info.ch_names)
+    if adjacency.shape[0] != n_channels:
+        raise ValueError('``adjacency`` must have the same number of rows '
+                         'as the number of channels in ``inst``. Found '
```
```suggestion
                         'as the number of channels in ``info``. Found '
```

mmagnuski commented:
I also get this on CIs:
```
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.10.4/x64/bin/mne", line 33, in <module>
    sys.exit(load_entry_point('mne==1.1.dev0', 'console_scripts', 'mne')())
  File "/opt/hostedtoolcache/Python/3.10.4/x64/lib/python3.10/site-packages/mne-1.1.dev0-py3.10.egg/mne/commands/utils.py", line 107, in main
    cmd.run()
  File "/opt/hostedtoolcache/Python/3.10.4/x64/lib/python3.10/site-packages/mne-1.1.dev0-py3.10.egg/mne/commands/mne_sys_info.py", line [32](https://github.com/mmagnuski/mne-python/runs/6530498348?check_suite_focus=true#step:8:33), in run
    mne.sys_info(show_paths=options.show_paths, dependencies=dependencies)
  File "/opt/hostedtoolcache/Python/3.10.4/x64/lib/python3.10/site-packages/mne-1.1.dev0-py3.10.egg/mne/utils/config.py", line 616, in sys_info
    out(f' {{backend={mod.get_backend()}}}')
  File "/opt/hostedtoolcache/Python/3.10.4/x64/lib/python3.10/site-packages/matplotlib/__init__.py", line 1170, in get_backend
    return rcParams['backend']
  File "/opt/hostedtoolcache/Python/3.10.4/x64/lib/python3.10/site-packages/matplotlib/__init__.py", line 672, in __getitem__
    plt.switch_backend(rcsetup._auto_backend_sentinel)
  File "/opt/hostedtoolcache/Python/3.10.4/x64/lib/python3.10/site-packages/matplotlib/pyplot.py", line 268, in switch_backend
    switch_backend(candidate)
  File "/opt/hostedtoolcache/Python/3.10.4/x64/lib/python3.10/site-packages/matplotlib/pyplot.py", line 288, in switch_backend
    class backend_mod(matplotlib.backend_bases._Backend):
  File "/opt/hostedtoolcache/Python/3.10.4/x64/lib/python3.10/site-packages/matplotlib/pyplot.py", line 289, in backend_mod
    locals().update(vars(importlib.import_module(backend_name)))
  File "/opt/hostedtoolcache/Python/3.10.4/x64/lib/python3.10/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/opt/hostedtoolcache/Python/3.10.4/x64/lib/python3.10/site-packages/matplotlib/backends/backend_qtagg.py", line 9, in <module>
    from .qt_compat import QT_API, _enum, _setDevicePixelRatio
  File "/opt/hostedtoolcache/Python/3.10.4/x64/lib/python3.10/site-packages/matplotlib/backends/qt_compat.py", line 1[37](https://github.com/mmagnuski/mne-python/runs/6530498348?check_suite_focus=true#step:8:38), in <module>
    _setup()
  File "/opt/hostedtoolcache/Python/3.10.4/x64/lib/python3.10/site-packages/matplotlib/backends/qt_compat.py", line 91, in _setup_pyqt5plus
    from PyQt6 import QtCore, QtGui, QtWidgets, sip
RuntimeError: the sip module implements ABI v13.0 to v13.3 but the PyQt6.QtCore module requires ABI v13.4
matplotlib:          3.5.2
Error: Process completed with exit code 1.
```

mmagnuski commented:
Also, the tutorial fails because MEG names are different between adjacency and read epochs file:
![image](https://user-images.githubusercontent.com/8452354/169607128-1b6df72f-5d21-4bf7-b304-63387014d823.png)

The read adjacency uses the following names:
```
['MEG0111',
 'MEG0121',
 'MEG0131',
...
```
while the epochs file contains such `.ch_names`:
```
['MEG 0111',
 'MEG 0121',
 'MEG 0131',
...
```
I'm not sure what's the preferred fix here:
* automatically change channel naming to mne python convention when reading FieldTrip adjacency
* fix the adjacency we distribute with mne

sappelhoff commented:
I think I'd prefer 

> automatically change channel naming to mne python convention when reading FieldTrip adjacency

mmagnuski commented:
I'd also prefer this but I'm not sure if the pattern I see in this case is universal. Do we always name these MEG channels including the space?

sappelhoff commented:
This is a problem for the neuromag sensors apparently. We have two neighbor files for these:

- neuromag306mag_neighb.mat
- neuromag306planar_neighb.mat

for both of them, the names are organized as:

```matlab
>> load /home/stefanappelhoff/Desktop/bids/mne-python/mne/channels/data/neighbors/neuromag306mag_neighb.mat
>> neighbours.label
ans = MEG0111
ans = MEG0121
ans = MEG0131
ans = MEG0141
ans = MEG0211
ans = MEG0221
```
```matlab
>> load /home/stefanappelhoff/Desktop/bids/mne-python/mne/channels/data/neighbors/neuromag306planar_neighb.mat
>> neighbours.label
ans = MEG0113
ans = MEG0112
ans = MEG0122
ans = MEG0123
ans = MEG0132
ans = MEG0133
ans = MEG0143
ans = MEG0142
ans = MEG0213
ans = MEG0212
```

if MNE-Python uses the same names but with a space between `MEG` and the 4-digit number, then I think you could for now hardcode the space in `read_ch_adjacency` so that the neighbor template names and the epochs names are the same.

As for the other templates: Maybe they are not a problem, let's cross that bridge when we reach it. We don't even have **that** many templates, and for EEG, these are not used by default.

Wdyt @larsoner?

mmagnuski commented:
Ok, I can do a small fix now and check the FieldTrip neighbour conventions vs mne in a next PR to see if we can solve it more generally.

mmagnuski commented:
## Potentially last step - enhancing the visuals

This is how the adjacency looks for this MEG template:
![image](https://user-images.githubusercontent.com/8452354/169644920-c983c844-86c4-4416-ba7f-04f8b9413ed1.png)

Let me know if you think the visuals should be tweaked.
I think for example that the head outline gets a bit lost among these lines.

Changing the adjacency color makes it a little better, I think:
![image](https://user-images.githubusercontent.com/8452354/169645063-0f964bd5-beab-40c0-ba38-88d75bb43976.png)

Another option is to also change the adjacency line width:
![image](https://user-images.githubusercontent.com/8452354/169645195-3d62b9fc-659f-4f90-92ff-c79dbdbf9443.png)

hoechenberger commented:
I like the adjustment of the line width. Maybe adding a little bit of transparency to the nodes (alpha 0.6 or so) could help too?

sappelhoff commented:
wow this looks super good! I was expecting it to be much more messy :-)

> I like the adjustment of the line width. 

+1

mmagnuski commented:
> Maybe adding a little bit of transparency to the nodes (alpha 0.6 or so) could help too?

Good idea, I'll test that.

mmagnuski commented:
adding alpha of 0.7 to the nodes and removing their linewidth:
![image](https://user-images.githubusercontent.com/8452354/169646436-627ac2cf-0ebe-4911-82e6-11ca0c730515.png)

hoechenberger commented:
I have a general question as to the definition of adjacency here. In the following figure (taken from your screenshot above), I totally see how the green edges point to neighbors of the green node. But I think one could debate whether the nodes connected via the red edges are neighbors of the same kind, no? Shouldn't at least some sort of (reduced) weight be applied here? Or am I just getting confused as we're looking at a 2D projection of a 3D mesh?

![image](https://user-images.githubusercontent.com/2046265/169646727-960ff3bd-0395-436b-9eb7-9867ea063ac2.jpeg)

hoechenberger commented:
BTW the viz looks great!

mmagnuski commented:
I think this is a case where projecting to 2d just looks weird, but I'm working on the 3d mode now, so I'll post a screenshot soon (BTW plotting sensors from info with `mne.viz.plot_sensors(info, ...)` and using the pos field of `info['chs']` gave different 3d positions, may be related to different coordinates (head vs not head)? I felt it may be another pandora box so I use the plotted point positions in 3d now).

mmagnuski commented:
but I agree that the red connections could be removed :)

mmagnuski commented:
I think I'll disable node scaling in 3d mode, I get weird matplotlib errors:

```python-traceback
TypeError: only integer scalar arrays can be converted to a scalar index

Traceback (most recent call last):

  File "C:\Users\Admin\.conda\envs\mne\lib\site-packages\IPython\core\formatters.py", line 341, in __call__
    return printer(obj)

  File "C:\Users\Admin\.conda\envs\mne\lib\site-packages\IPython\core\pylabtools.py", line 151, in print_figure
    fig.canvas.print_figure(bytes_io, **kw)

  File "C:\Users\Admin\.conda\envs\mne\lib\site-packages\matplotlib\backend_bases.py", line 2295, in print_figure
    self.figure.draw(renderer)

  File "C:\Users\Admin\.conda\envs\mne\lib\site-packages\matplotlib\artist.py", line 73, in draw_wrapper
    result = draw(artist, renderer, *args, **kwargs)

  File "C:\Users\Admin\.conda\envs\mne\lib\site-packages\matplotlib\artist.py", line 50, in draw_wrapper
    return draw(artist, renderer)

  File "C:\Users\Admin\.conda\envs\mne\lib\site-packages\matplotlib\figure.py", line 2810, in draw
    mimage._draw_list_compositing_images(

  File "C:\Users\Admin\.conda\envs\mne\lib\site-packages\matplotlib\image.py", line 132, in _draw_list_compositing_images
    a.draw(renderer)

  File "C:\Users\Admin\.conda\envs\mne\lib\site-packages\matplotlib\artist.py", line 50, in draw_wrapper
    return draw(artist, renderer)

  File "C:\Users\Admin\.conda\envs\mne\lib\site-packages\mpl_toolkits\mplot3d\axes3d.py", line 451, in draw
    for artist in sorted(collections_and_patches,

  File "C:\Users\Admin\.conda\envs\mne\lib\site-packages\mpl_toolkits\mplot3d\axes3d.py", line 440, in do_3d_projection
    return artist.do_3d_projection()

  File "C:\Users\Admin\.conda\envs\mne\lib\site-packages\matplotlib\_api\deprecation.py", line 386, in wrapper
    return func(*inner_args, **inner_kwargs)

  File "C:\Users\Admin\.conda\envs\mne\lib\site-packages\mpl_toolkits\mplot3d\art3d.py", line 614, in do_3d_projection
    self._sizes = self._sizes3d[z_markers_idx]

TypeError: only integer scalar arrays can be converted to a scalar index

<Figure size 432x432 with 1 Axes>
```

although you can see `z_markers_idx` is an array of int...
```python
%debug
> c:\users\admin\.conda\envs\mne\lib\site-packages\mpl_toolkits\mplot3d\art3d.py(614)do_3d_projection()
    612 
    613         if len(self._sizes3d) > 1:
--> 614             self._sizes = self._sizes3d[z_markers_idx]
    615 
    616         if len(self._linewidths3d) > 1:


z_markers_idx
[ 81  80  79  65  72  88  78  97  71  89  64  62  77  94  73  76  96  87
  70  86  63  74  56  61  75  95  60  92 100  69  93  84  55  68  58  57
  85 101  59  91  99  66  90  67  54  83  82  27   7   3  26  98   6  48
  52  15  49   2  14  40  41   4  24   0  25  53  47   5  51  12   1  46
  13  39  38  22  11  37  50  43   9  10  44  23  45  36  21   8  42  19
  20  35  32  34  16  31  18  33  29  17  30  28]

type(z_markers_idx)
<class 'numpy.ndarray'>

z_markers_idx.dtype
int64
```

mmagnuski commented:
![image](https://user-images.githubusercontent.com/8452354/169647899-ea440c46-b23a-435a-85be-891dd5ea02be.png)

but not changing the alpha and linewidths of the points in 3d allows matplotlib to change it dynamicaly based on perspecitve and the depth is easier to understand then:
![image](https://user-images.githubusercontent.com/8452354/169648042-1e707db6-8587-4578-8302-9616af1f5e5e.png)

so I think I'll set the alpha and linewidths for the nodes only in 2d.

mmagnuski commented:
another one in 3d without linewidths:
![image](https://user-images.githubusercontent.com/8452354/169648129-e933dd5a-3e2d-4933-8a77-ba4a13e5ef0b.png)

sappelhoff commented:
I'd say for 3d the dynamic adjustment based on perspective is preferrable --- but for 2d it's nice with alpha and line-width adjustments hardcoded

> but I agree that the red connections could be removed :)

wouldn't that rather be a different PR that is having a go at the delaunay triangulations?

mmagnuski commented:
> wouldn't that rather be a different PR that is having a go at the delaunay triangulations?

Oh, yes, I wouldn't change the template. I think the template is very good in general and these red connections may be a matter of preferrence and one could change them manually with `edit=True`.

> I'd say for 3d the dynamic adjustment based on perspective is preferrable --- but for 2d it's nice with alpha and line-width adjustments hardcoded

Great, I'll leave it for 2d, that's for sure. But I'm not certain which one to pick for 3d: linewidth or no linewidths? :) (the last two plots, one in a separate comment)

sappelhoff commented:
I like it better when the markers don't have additional lines :-)

mmagnuski commented:
also @hoechenberger:
> Shouldn't at least some sort of (reduced) weight be applied here? 

That is not possible in conventional clustering - channels either are neighbours or not. But one can reasonably remove these connections now with `edit=True`. Generally, sticking with the templates that FieldTrip distributes is good for cross-package consistency, but these templates may change from time to time on FT side (that's what they say in their docs), so we could check if they are still the same.

sappelhoff commented:
> so we could check if they are still the same.

I had a look, they were last updated 10 years ago: https://github.com/fieldtrip/fieldtrip/tree/master/template/neighbours

mmagnuski commented:
I've added a see also section, but I should do something additional to make the rendered docstring use links for these function names (the same way it looks [here in the "see also" section](https://mne.tools/dev/generated/mne.channels.find_ch_adjacency.html)). Do you have a hint?

mmagnuski commented:
ok, I think I know :)

mmagnuski commented on mne/viz/topomap.py:
```diff
@@ -2806,3 +2807,206 @@ def plot_bridged_electrodes(info, bridged_idx, ed_matrix, title=None,
+                            for i in range(adjacency.shape[0])]
+        node_size = [max(x, 3) ** 2.5 for x in n_connections]
+        path_collection[0].set_sizes(node_size)
+    else:
```
how is that `E127 continuation line over-indented for visual indent`?

mmagnuski replied:
ah, its a few lines higher...

mmagnuski commented:
https://output.circle-artifacts.com/output/job/e63439a6-50e9-40eb-a1ca-81423ea379c6/artifacts/0/dev/auto_tutorials/stats-sensor-space/75_cluster_ftest_spatiotemporal.html#find-the-fieldtrip-neighbor-definition-to-setup-sensor-adjacency

sappelhoff reviewed (approved):
wonderful! +1 to merge once CIs are green. Thanks for this awesome addition

drammock commented:
This is a really great addition, thanks @mmagnuski and @sappelhoff !

agramfort commented on mne/viz/topomap.py:
```diff
@@ -2806,3 +2807,206 @@ def plot_bridged_electrodes(info, bridged_idx, ed_matrix, title=None,
+    adjacency : array
+        Array of channels x channels shape. Defines which channels are adjacent
+        to each other. Note that if you edit adjacencies
+        (via ``edit=True``), this array will be modified in place.
```
requiring a numpy dense array in edit mode is I think an unecessary.
Yes inserting elements in a sparse array is inefficient (at least for CSR,  CSC formats) but it should not be critical for our size of problem. I would do something like this:

```
from scipy import sparse
import warnings

A = sparse.random(10, 10, density=0.1).tocsr()
with warnings.catch_warnings(record=True):
    A[0, 0] = 1
```

mmagnuski replied:
Ok, I think I can rework the code for `csr_matrix`. So I undrestand that `coo_matrix` should never appear as the adjacency (I think this sparse format was used previously in mne). If so a check for either dense or csr (and otherwise error) would be useful.

sappelhoff replied:
True, it would be nice to somehow solve this problem ... otherwise people will always write three lines:

```python
adj = find_ch_adjacency
adj = adj.to_array()  # <-- would be cool if we could get rid of this
plot_ch_adjacency(adj)
```

mmagnuski replied:
plotting and editing sparse arrays is now possible, it was easier than I thought. :)

agramfort commented on mne/viz/topomap.py:
```diff
@@ -2806,3 +2807,206 @@ def plot_bridged_electrodes(info, bridged_idx, ed_matrix, title=None,
+        channel (depending on whether the channels are already adjacent or
+        not); the newly clicked channel now becomes activated. Clicking on
+        an activated channel deactivates it. Editing is currently only
+        supported for ``kind='2d'``.
```
when you click outside of any electrode i think it should remove the green segments

mmagnuski replied:
I can change it, but it may be a bit frustrating if someone misses the desired channel (to add/remove connection) from time to time and it deselects his current one.

sappelhoff replied:
@agramfort did you pull the changes and try it yourself? I find the user experience of the current interface quite nice.

mmagnuski replied:
I'm also not sure how it should be implemented. Currently only pick events are handled. To support clicking outside I'd have to add another mouse event handler. But I think then if one clicks on the channel both events would fire, so the second mouse event would have to somehow detect if the first event was activated?

sappelhoff replied:
I think we should keep the current user interface -- if trying the code in real life shows us that indeed we should change it because there are pain points, let's do a follow up PR.

sappelhoff commented:
Thanks @mmagnuski :tada: :rocket:

mmagnuski commented:
Thanks @sappelhoff @hoechenberger @agramfort ! 🚀
